Multigrid support [drzisga/multigrid]

drzisga opened the pull request:
This pull request adds general matrix-based and matrix-free multigrid support. Currently, only H1 spaces are supported.

- [x] This PR relies heavily on #1107 and #982 which should be merged first.

<!--GHEX{"id":1097,"author":"drzisga","editor":"tzanio","reviewers":["delyank","barker29","pazner"],"assignment":"2020-02-09","approval":"2020-04-06","merge":"2020-04-13T07:00:00.000Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1097](https://github.com/mfem/mfem/pull/1097) | @drzisga | @tzanio | @delyank + @barker29 + @pazner | 02/09/20 | 04/06/20 | ⌛due 04/13/20 | |
<!--ELBATXEHG-->

barker29 commented on fem/spacehierarchy.hpp:
```diff
@@ -0,0 +1,94 @@
+
+#ifdef MFEM_USE_MPI
+using ParSpaceHierarchy =
+    AbstractSpaceHierarchy<ParMesh, ParFiniteElementSpace>;
```
Since `Mesh` is a `ParMesh` and `FiniteElementSpace` is a `ParFiniteElementSpace`, I don't think this is necessary. Everything should just work fine with what you call `SpaceHierarchy` without any templates.

drzisga replied:
See my response to your other comment in this file for why I am using templates here.

barker29 commented on fem/transfer.cpp:
```diff
@@ -0,0 +1,188 @@
+namespace mfem
+{
+
+TransferOperator::TransferOperator(const FiniteElementSpace &lFESpace_,
```
This is a nice abstraction.

drzisga replied:
Thank you.

barker29 commented on linalg/solvers.cpp:
```diff
@@ -146,6 +146,134 @@ void OperatorJacobiSmoother::Mult(const Vector& x, Vector &y) const
+   MFEM_FORALL(i, ess_tdof_list.Size(), X[I[i]] = 1.0; );
+
+   // Set up Chebyshev coefficients
+   // See Python script for code generation
```
Maybe better to have some kind of reference/citation here?

drzisga replied:
I added a reference to a paper. Is there any place where we can put additional scripts which are useful for MFEM? I wrote a Python script for generating the code for the coefficients.

barker29 commented:
One more comment: the code that allows variable coefficients in the PA kernels could be separated into a separate PR. That functionality is useful even apart from the multigrid stuff, and it might get a faster review by itself.

drzisga commented on fem/bilininteg_diffusion.cpp:
```diff
@@ -105,6 +105,37 @@ static void PADiffusionSetup2D(const int Q1D,
 }
 
+// PA variable coefficient Diffusion Assemble 2D kernel
+static void PAVarCoeffDiffusionSetup2D(const int Q1D,
```
I wanted to do this first, but then the OCCA kernels need to be changed too, which I don't know how to test yet. I will create an extra pull request in this regard.

camierjs replied:
I can take a look at these coeffs, I need it too for Laghos.

drzisga replied:
You committed just shortly before I created the new pull request. Please see #1107

drzisga commented on fem/spacehierarchy.hpp:
```diff
@@ -0,0 +1,94 @@
+{
+
+/// Class bundling a hierarchy of meshes and finite element spaces
+template <typename M, typename FES> class AbstractSpaceHierarchy
```
This was my initial idea idea, but it makes the class much more complicated for users to use. For instance, if a user wants to create a new ParBilinearForm using a FiniteElementSpace returned by the SpaceHierarchy, he needs to cast it manually to a ParFiniteElementSpace. Another issue arises in the AddUniformlyRefinedLevel method where a new FiniteElementSpace is being allocated and the required type can (for my understanding) not be deducted at run time.

pazner replied:
One way of checking the type is performing a `dynamic_cast` to e.g. `ParFiniteElementSpace *` and checking for `nullptr`

barker29 replied:
In addition to using Will's suggestion in `AddUniformlyRefinedLevel`, I suggest removing all the GetMesh methods (the user can do `GetFESpace().GetMesh()`), and have a `ParSpaceHierarchy` object that overloads only the GetFESpace() type methods.

There are other options, but I think this would fit into existing MFEM code the best. We tend to avoid templates (whether or not that's a good decision, but it's a decision that's already been made).

drzisga replied:
I rewrote the SpaceHierarchy using polymorphism and overloading so it fits better into the current MFEM code.

drzisga commented on fem/coefficient.hpp:
```diff
@@ -156,6 +156,8 @@ class FunctionCoefficient : public Coefficient
    virtual double Eval(ElementTransformation &T,
                        const IntegrationPoint &ip);
+
+   double Eval(const Vector& x);
```
You are right, it is confusing. I will rename it and add some comment for documentation.

drzisga commented on linalg/multigrid.hpp:
```diff
@@ -0,0 +1,183 @@
+   /// the call to this method additional levels may be added with AddLevel. The
+   /// ownership of the operators and solvers may be transferred to the
+   /// MultigridOperator by setting the according boolean variables.
+   void AddCoarseLevel(Operator* opr, Solver* solver, bool ownOperator,
```
I renamed this method.

drzisga commented on fem/transfer.hpp:
```diff
@@ -0,0 +1,64 @@
+namespace mfem
+{
+
+class TransferOperator : public Operator
```
I missed this, thank you for noticing.

drzisga commented on miniapps/performance/mg.cpp:
```diff
@@ -0,0 +1,569 @@
+#include "mfem.hpp"
```
I was using the templated code first, that is why I put the file there. Do you have any suggestion where to put the file instead?

pazner replied:
I actually think it's a pretty good location for this example. The `performance` directory doesn't actually have the templated code (those are all the `t*.?pp` files in e.g. the `fem` directory), it's just an example that uses the templated code to do some performance testing.

barker29 replied:
Fair enough. My initial thought was this solver would go in the regular ex1/ex1p example, but we're probably not ready for that yet.

drzisga commented on fem/fem.hpp:
```diff
@@ -31,6 +31,8 @@
 #include "staticcond.hpp"
 #include "tmop.hpp"
+#include "transfer.hpp"
+#include "spacehierarchy.hpp"
```
When the user wants to use these classes and only includes the mfem.hpp header in the cpp file, these includes need to be there. I understood the fem.hpp header to be some kind of collection of all headers in the fem directory or am I wrong?

barker29 replied:
You're right, I misunderstood which file I was looking at.

drzisga commented on linalg/multigrid.hpp:
```diff
@@ -0,0 +1,183 @@
+};
+
+// Multigrid solver
+class MultigridSolver : public Solver
```
Thank you.

pazner commented on fem/spacehierarchy.cpp:
```diff
@@ -0,0 +1,149 @@
+M& AbstractSpaceHierarchy<M, FES>::GetMeshAtLevel(unsigned level)
+{
+   return const_cast<M&>(
+       const_cast<const AbstractSpaceHierarchy*>(this)->GetMeshAtLevel(level));
```
I think it would be nicer to avoid using `const_cast` here

drzisga replied:
You are right, it is ugly and unnecessary here. I changed this while rewriting the SpaceHierarchy as discussed in the other comment.

drzisga commented on linalg/multigrid.hpp:
```diff
@@ -0,0 +1,183 @@
+{
+
+/// Abstract multigrid operator
+class MultigridOperator : public Operator
```
The class container itself should already support this feature, but there is currently no way to construct such an operator. I am thinking of good way to do this, but as you said, this is currently not of high priority.

drzisga commented on fem/occa.okl:
```diff
@@ -44,58 +44,64 @@ typedef double* SymmOperator3D_t @dim(Q3D, 6, NE);
-      op(q, 2, e) =  c_detJ * (J11*J11 + J12*J12); // (2,2)
-    }
-  }
+                              @restrict const DofToQuad_t C,
```
I do not know much about OCCA but it looks to me that the dimensions of C are wrong. This is also the case for the 3D setup. Please see the occa.okl file in #1107 for a comparison.

camierjs replied:
You're right, good catch!

drzisga commented on fem/transfer.hpp:
```diff
@@ -0,0 +1,112 @@
+};
+
+/// Matrix-free transfer operator between finite element spaces on the same mesh
+class OrderTransferOperator : public Operator
```
I changed the name

tzanio commented:
@drzisga and @v-dobrev -- is this still a WIP?

drzisga commented:
@v-dobrev still wants to take a look at it. This PR implements matrix-free methods only, but it can easily be extended to support matrix-based multigrid. I am not so happy about the (Par)MultigridBilinearForm classes, but it was the cleanest way to support boundary conditions in the MultigridOperator for me. This may need some more discussion.

Additionally, I added the p-Multigrid preconditoner to the partial assembly case in ex1 and ex1p which makes these examples more complicated for first users. The main miniapp for comparing the p-Multigrid to the preconditioners based on LOR is located in the performance/mg.cpp file. It is still not clear if this the right location for it.

This PR relies heavily on #1107 and #982 which should be merged first.

tzanio commented:
Given that #1107 and #982 will be both be merged soon, can we prepare this for review?

drzisga commented:
I am preparing it for review, but I have some questions left. Do you want me to keep the multigrid preconditioner in examples ex1 and ex1p? It makes them quite complicated as first examples. The only other usage of the multigrid preconditioner is in the performance/mg miniapp. Is this the preferred location or should I move it somewhere else?

tzanio commented:
I'd like to hear what other folks thing, but here is my take...

> Do you want me to keep the multigrid preconditioner in examples ex1 and ex1p?

No, we do want to keep Example 1 fairly simple.

One option is to add them as Example 23. The example codes are meant to be compared, so it OK if ex1 and ex23 are fairly similar.

> The only other usage of the multigrid preconditioner is in the performance/mg miniapp. Is this the preferred location or should I move it somewhere else?

This depends on what mg.cpp is intended to do. If it is more of an illustration, then it should be an example code. If it is intended to demonstrate high performance, then miniapps/performance is probably the best place for it. So far we have used that directory for the templated codes (as also reflected in the README there) but maybe this is a good time to expand it to other high-performance codes?

drzisga commented:
The PR should be ready for review now. I decided to add a new example 23 demonstrating the creation of a matrix-free solver with a p-multigrid preconditioner instead of keeping the miniapp. Please let me know what you think about that, so I can open a PR for the example in the web repository.

tzanio commented:
@delyank + @barker29 + @pazner -- any chance we can review this in the next week, or should we keep the current schedule and bump it to mfem-4.2?

pazner commented:
I will definitely look at this next week, but it's a big PR so I'm not sure it will be completely done by the end of the week.

tzanio commented:
Ok, lets just move it to mfem-4.2. If it happens early, that's great but no pressure.

pazner commented on fem/pmgbilinearform.hpp:
```diff
@@ -0,0 +1,44 @@
+namespace mfem
+{
+
+class ParMultigridBilinearForm : public MultigridBilinearForm
```
The same comments from `MultigridBilinearForm` also apply to this class.

It would also be great to avoid the code duplication associated with the `Par` objects. I seem to remember that you originally had a templated version, and changed to this non-templated version to make the code more in line with the MFEM style.

I don't have a great solution for this other than templates (perhaps we should consider using templates in this case), but I do think that the code duplication is an anti-pattern, which will make code maintenance more difficult and increase the likelihood of bugs.

drzisga replied:
You are right. These and the `SpaceHierarchy` classes were originally implemented using templates. Does anyone else have a preference of what to use in this case?

v-dobrev replied:
We can have one object for serial and parallel. If using just virtual methods from `FiniteElementSpace` or `BilinearForm` there will be no need to write different code in serial and parallel. If we need to distinguish the serial/parallel case we can use `dynamic_cast`, e.g. from `FiniteElementSpace*` to `ParFiniteElementSpace*`. This approach is used e.g. in class `StaticCondensation`.

delyank replied:
I'm with Veselin on this.

drzisga replied:
Please see my previous comments. I got rid of the ParMultigridBilinearForm and moved it to the examples.

pazner commented on fem/transfer.hpp:
Newline

drzisga commented on tests/unit/fem/test_transfer.cpp:
```diff
@@ -0,0 +1,100 @@
+         }
+      }
+   }
+}
```
Thanks for pointing this out. I added a newline at the end of this and all the other files.

drzisga commented on linalg/multigrid.cpp:
```diff
@@ -0,0 +1,419 @@
+
+MultigridOperator::~MultigridOperator()
+{
+   for (int i = operators.Size() - 1; i >= 0; --i)
```
I wanted to make sure that everything gets deleted in the reverse order of its creation. But this should be unnecessary here, so I reversed it again.

drzisga commented on linalg/multigrid.hpp:
```diff
@@ -0,0 +1,224 @@
+};
+
+/// Abstract multigrid operator with timing support
+class TimedMultigridOperator : public MultigridOperator
```
I removed this class for now and I am considering to re-add this in a miniapp.

drzisga commented on fem/bilininteg.hpp:
```diff
@@ -173,6 +173,13 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
+      MFEM_ABORT("Copy not implemented ");
+      return nullptr;
+   }
+
```
I agree with you. I don't like this solution so much, but it was required for the implementation of the MultigridBilinearForm classes. Please see my comment there.

drzisga replied:
I decided to get rid of the Copy methods and moved the MultigridBilinearForm classes into the examples.

drzisga commented on fem/mgbilinearform.hpp:
```diff
@@ -0,0 +1,43 @@
+namespace mfem
+{
+
+class MultigridBilinearForm : public MultigridOperator
```
The reason I decided to implement a `MultigridBilinearForm` was that the user can easily create a new `MultigridOperator` with sane default smoothers and a reasonable coarse grid solver. This class also handles the memory of the forms which are required for each level in the hierarchy. In this way, the user only needs to define a bilinear form on the finest level and to provide a `SpaceHierarchy`. The bilinear forms for the other levels are being created automatically for all spaces in the `SpaceHierarchy` (that is what I need the newly introduced copy method in the integrators for).

I discussed with Veselin once, that we want the `MultigridOperator` class to be a linear algebra object which should not "know" anything about BilinearForms. This would be violated when adding the "named constructor" to `MultigridOperator`.

Do you think it would be okay to move what's inside of `MultigridBilinearForm` to the example file? Then the user has to do more work but can choose between different smoothers and coarse grid solvers.

delyank replied:
I might be the one of the few who actually likes the idea of the `MultigridBilinearForm`. I agree with Veselin's perspective that the operators should be agnostic about specifics like bilinear forms.

If there is projected generic future use of this class, it's better to remain in the library. If not, the compromise to move it to the example seems reasonable. If in the future it becomes necessary it can be pulled out of the example back to the library.

drzisga replied:
I decided to move the (Par)MultigridBilinearForm classes into the respective examples. In the current state, these classes were too inflexible for the general library, as they are too application specific. I do think, however, that adding such class in the future may be reasonable.

drzisga commented on fem/mgbilinearform.cpp:
```diff
@@ -0,0 +1,128 @@
+   for (int i = 0; i < dbfi.Size(); ++i)
+   {
+      form->AddDomainIntegrator((*bf.GetDBFI())[i]->Copy());
+   }
```
Thank you, I did not see the `seExternalIntegrators()` method. I will try this out instead of using my `Copy()` methods.

barker29 replied:
I strongly agree. I don't like the `Copy()` at all.

drzisga replied:
I remember now why I added the `Copy` methods. I actually tried it with `UseExternalIntegrators` before but that does not work for partially assembled integrators. These integrators have their own member variables which store the pre-computed values. The pre-computed values depend on the finite element space and can not be re-used for finite element spaces of different orders (in the multigrid hierarchy). This means, the integrators cannot be re-used with `UseExternalIntegrators` since they have their finite element space dependent state. I agree with you and do not like the `Copy` methods myself, but I do not see another solution for this issue right now.

v-dobrev replied:
This is a good point. There are ways to avoid the use of `Copy` but they will probably require more changes than just adding `Copy`. There is a similar method for class `Element` called `Duplicate`; another possible name is `Clone`.

delyank replied:
This is actually confusing. If you copy the "skeleton" of the integrator but not the particular data it may contain, then should you call it "copy", "clone", or "duplicate". All those seem confusing. This is a form of a "shallow copy" or even "transfer" of the integrator, or "recreate". This is likely to be something that would bite back at us in the future.

Maybe it's too late but one way would have been to have "context"/"state" of the integrator stored separately from the integrator and invoking the integrator with the provision of the context. Then, it would be a context switch every time you reuse it, as it is a common way to use a shared resource.

drzisga replied:
You are right @delyank. The name was very confusing. I decided to move the (Par)MultigridBilinearForm classes to the examples and to get rid of the Copy methods. Independently of this pull request, it might be necessary to add these methods again in the future, but this should be done in a separate pull request/discussion.

drzisga commented:
Thank you for the review @pazner. It is very hard for me to find the right balance between simplicity and generality. My current approach was to make it as easy as possible for the user to create a `MultigridOperator` without the hassle of handling all the bilinear forms for the spaces in the hierarchy. We could get rid of that and let the users create their own application specific `MultigridOperator` and give them the possibility of choosing the smoothers and coarse grid solver. This approach would take away some complexity for the developers. Is there some approach you (and the others) prefer?

barker29 commented on fem/mgbilinearform.hpp:
```diff
@@ -0,0 +1,43 @@
+
+public:
+   /// Empty constructor
+   MultigridBilinearForm();
```
I would remove this. It isn't used and I don't think it makes sense.

drzisga replied:
Done.

barker29 commented on fem/pmgbilinearform.cpp:
```diff
@@ -0,0 +1,126 @@
+
+   HypreBoomerAMG* amg = new HypreBoomerAMG(*hypreCoarseMat);
+   amg->SetPrintLevel(-1);
+   amg->SetMaxIter(1);
```
This is the default, so I would remove this line. (We might not even want to add the `SetMaxIter()` method in this PR since this is the only place it's used.)

drzisga replied:
I reverted these changes.

barker29 commented on fem/transfer.hpp:
```diff
@@ -0,0 +1,158 @@
+
+private:
+   /// Helper function to decode encoded dofs
+   static inline int DecodeDof(int dof)
```
Recommend using `lFESpace.DecodeDof()` and removing this implementation.

drzisga replied:
I removed this implementation and use the one from `FiniteElementSpace` now.

barker29 commented on examples/ex23.cpp:
```diff
@@ -0,0 +1,209 @@
+   {
+      collections.Append(new H1_FECollection(orders[level], dim));
+      FiniteElementSpace *fesp = new FiniteElementSpace(mesh, collections.Last());
+      spaceHierarchy.AddLevel(mesh, fesp, false, true);
```
I think it makes sense to demonstrate `SpaceHierarchy::AddOrderRefinedLevel()` here in the example.

drzisga replied:
Thank you, I totally forgot about that method myself. I updated the examples to use this method.

barker29 commented on examples/ex23p.cpp:
```diff
@@ -0,0 +1,264 @@
+      collections->Append(new H1_FECollection(orders[level], dim));
+      ParFiniteElementSpace* fesp =
+         new ParFiniteElementSpace(pmesh, collections->Last());
+      spaceHierarchy->AddLevel(mesh, fesp, false, true);
```
`AddOrderRefinedLevel()`

barker29 commented:
My PR is a bit garbled because of [this](https://www.githubstatus.com/incidents/xp2qc958g4wt) but I think all the comments are there.

tzanio commented:
Pushing the review for this PR for after mfem-4.1

delyank commented on fem/mgbilinearform.cpp:
```diff
@@ -0,0 +1,128 @@
+   form->FormSystemMatrix(*essentialTrueDofs.Last(), opr);
+   opr.SetOperatorOwner(false);
+
+   CGSolver* pcg = new CGSolver();
```
Sometimes creating too many examples for each and every functionality might be burdensome and generate too much code. It is always nice to have examples, but just documenting things in more detail can replace having way to detailed examples.

drzisga replied:
You are right. I moved this to the examples now. There it is shown how to construct a `MultigridOperator` from scratch. The user can then decide which solvers or smoothers to use.

delyank commented on linalg/hypre.cpp:
```diff
@@ -2779,6 +2779,11 @@ void HypreBoomerAMG::SetDefaultOptions()
+{
+   HYPRE_BoomerAMGSetMaxIter(amg_precond, maxIter);
+}
+
```
I think that one of the philosophies of MFEM is to do the iterative processes within MFEM constructs while preconditioners may come from outside. This can probably be useful if for example you want to precondition by 2 Vcycles. However, that can also be resolve within what MFEM already provides.

I'm actually negative towards this addition, because it would allow too much unregulated freedom.

drzisga replied:
I reverted these changes.

delyank commented on linalg/solvers.hpp:
```diff
@@ -115,6 +115,44 @@ class OperatorJacobiSmoother : public Solver
+       It is assumed the underlying operator acts as the identity
+       on entries in ess_tdof_list, corresponding to (assembled) DIAG_ONE
+       policy or ConstrainedOperator in the matrix-free setting. */
+   OperatorChebyshevSmoother(Operator* oper_, const Vector &d,
```
You can either run power method or do Gershgorin estimate (should be doable in matrix free (?))

drzisga replied:
I liked the idea of @barker29 and added another constructor which uses the `PowerMethod` internally to estimate the largest eigenvalue. @delyank do you mean to use the Gershgorin circle theorem? If yes, I do not think it is possible to get a tight estimate on the largest eigenvalue with it since you only obtain a set of intervals where the eigenvalues lie in. An alternative to the power method would be the Lanczos algorithm which can be used for symmetric matrices only but has better convergence properties.

delyank commented on fem/mgbilinearform.cpp:
```diff
@@ -0,0 +1,128 @@
+                                             int chebyshevOrder)
+   : MultigridOperator()
+{
+   MFEM_VERIFY(bf.GetAssemblyLevel() == AssemblyLevel::PARTIAL,
```
Yes, the pull request seems to promise matrix-free and matrix-assembled functionality. Why this restriction suddenly appear here?

drzisga replied:
The `SpaceHierarchy`, `MultigridOperator` and `MultigridSolver` work for any underlying operators. The `MultigridBilinearForm` was specific for matrix-free operators. This was another reason why I moved it to the examples.

delyank commented:
> @delyank + @barker29 + @pazner -- any chance we can review this in the next week, or should we keep the current schedule and bump it to mfem-4.2?

I'm generally positive about this pull request. It may introduce many classes but it's still reasonably neat. It mostly introduces new functionality. It also modifies some existing files and classes. Still, the modifications there are mostly new things and does not seem to be introducing danger as far as existing (or core) functionality is concerned.

I have some notes, mentioned in the code discussion, but they are mostly non-obligatory. I would prefer if there were more comments. The size of the pull request together with its dependence on other pull requests makes it somewhat harder to process.

I strongly dislike the `Copy` in the integrators but this is probably not an issue of this pull request but rather a reflection of an existing deficiency. The problem is that deficiency that are delayed to be addressed in the long run may grow into monsters.

drzisga commented on linalg/eigenvalue.hpp:
```diff
@@ -0,0 +1,88 @@
+#ifndef MFEM_EIGENVALUE
```
I added the license header to this file.

drzisga commented on fem/fespace.cpp:
```diff
@@ -1037,46 +1047,90 @@ ::Mult(const Vector &x, Vector &y) const
+      subY.SetSize(lP.Width());
 
-            if (!processed[r])
+      for (int vd = 0; vd < vdim; vd++)
```
I added tests covering the case of vector spaces in `test_transfer.cpp`

drzisga commented on linalg/multigrid.hpp:
```diff
@@ -0,0 +1,185 @@
+   virtual void InterpolateFrom(int level, const Vector& x, Vector& y) const;
+
+   /// Apply Smoother at given level
+   virtual void ApplySmootherAtLevel(int level, const Vector& x,
```
The smoother is closely related to the operator on each level and I wanted the `MultigridSolver` to depend only on a single operator object. If desired, I will move the smoothers from the `MultigridOperator` to their own `MultigridSmoother` class.

barker29 replied:
This is fine, I was just curious what the reasoning was.

drzisga commented on fem/spacehierarchy.hpp:
```diff
@@ -0,0 +1,118 @@
+   int GetNumLevels() const;
+
+   /// Returns the index of the finest level
+   int GetFinestLevelIndex() const;
```
From my previous work, I am used to 0 being the coarsest space and a larger level meaning a finer space. Is there a desired order in MFEM?

barker29 replied:
I prefer 0 to be finest but it really doesn't matter, as long as documentation/examples are clear.

drzisga commented on examples/ex23p.cpp:
```diff
@@ -0,0 +1,264 @@
+      spaceHierarchy->AddLevel(mesh, fesp, false, true);
+   }
+   spaceHierarchy->AddLevel(mesh, fespace, false, false);
+   mgOperator = new ParMultigridBilinearForm(*spaceHierarchy, *a, ess_bdr);
```
I agree and moved that code to the examples where the user can decide which coarse solvers/smoothers to use.

drzisga commented on fem/spacehierarchy.hpp:
```diff
@@ -0,0 +1,118 @@
+   /// Constructs an empty space hierarchy
+   SpaceHierarchy();
+
+   /// Constructs a space hierarchy with the given mesh and space on level zero.
```
I do not understand these comments, but I assume they are related to the next one?

barker29 replied:
```suggestion
   /// Constructs a space hierarchy with the given mesh and space on the coarsest level.
```

drzisga commented on CHANGELOG:
```diff
@@ -101,6 +102,10 @@ Discretization improvements
+- Added support for matrix-free interpolation and restriction operators between
+  finite element spaces of different order on the same mesh or with the same
+  order on uniformly refined meshes.
+
```
I have only tested them for continuous H1 spaces with simplicial and hexahedral elements, but I suppose they also work for discontinous (L2) spaces. These methods probably won't work for RT and ND spaces. I added "continuous H1 finite element spaces" to the CHANGELOG.

drzisga commented on examples/ex23p.cpp:
```diff
@@ -0,0 +1,264 @@
+//               -Delta u = 1 with homogeneous Dirichlet boundary conditions
+//               as in example 1. It highlights on the creation of a hierarchy
+//               of discretization spaces with partial assembly and the
+//               construction of an efficient p-multigrid preconditioner for the
```
I added a command line option to set the number of geometric h-refinements that are done prior to the order refinements.

tzanio commented:
Please resolve merge conflicts.

@pazner, do you have outstanding concerns?

pazner commented on examples/ex25p.cpp:
```diff
@@ -0,0 +1,329 @@
+   // 5. Refine the serial mesh on all processors to increase the resolution. In
+   //    this example we do 'ref_levels' of uniform refinement. We choose
+   //    'ref_levels' to be the largest number that gives a final mesh with no
+   //    more than 10,000 elements.
```
```suggestion
   //    more than 1,000 elements.
```

(there is a typo either in the comment or in the code — probably 1,000 is better for the serial refinements, since you do two parallel refinements later)

pazner commented:
I am seeing some strange behavior when running ex25p (default options).

1 MPI rank: 19 iterations
2 MPI ranks: 61 iterations
4 MPI ranks: 64 iterations
8 MPI ranks: 79 iterations
16 MPI ranks: 107 iterations
32 MPI ranks: 120 iterations

This seems like a bug to me. Any ideas?

psocratis commented:
> I am seeing some strange behavior when running ex25p (default options).
> 
> 1 MPI rank: 19 iterations
> 2 MPI ranks: 61 iterations
> 4 MPI ranks: 64 iterations
> 8 MPI ranks: 79 iterations
> 16 MPI ranks: 107 iterations
> 32 MPI ranks: 120 iterations
> 
> This seems like a bug to me. Any ideas?

It's possible that this is caused by HypreBoomerAMG/HypreGSSmoother because it depends on the number of ranks

pazner commented:
HypreBoomerAMG should only have an effect on the coarsest level, right? I tried changing the relaxation method to l1-scaled Jacobi (which shouldn't change with the number of ranks) and observed the same behavior.

psocratis commented:
> HypreBoomerAMG should only have an effect on the coarsest level, right? I tried changing the relaxation method to l1-scaled Jacobi (which shouldn't change with the number of ranks) and observed the same behavior.

Yes, only on the coarsest level i assume. Well then it's not that I guess!

drzisga commented:
Thank you for noticing this @pazner. I always observed some slight increase of iteration numbers when using more MPI ranks, but this seems too extreme for me. Another location which may change results in parallel, is the maximum eigenvalue estimation for the Chebyshev iteration. I will look into this and see if some older versions had the same issue.

tzanio commented:
I agree that something is wrong. If it is not the smoother, it could be the prolongation operators -- have those been verified in parallel?

drzisga commented:
You are right @tzanio. There is an issue with the transfer operators in parallel. I pushed a commit to make it work on CPUs again, but I think this change will break the transfers on GPUs. I haven't checked this yet, but I remember having a similar issue when I was at the lab. I will look more into this issue and will try to add some tests for the transfers in parallel.

tzanio commented:
Extending the review due date to address uncovered convergence issues...

drzisga commented:
I added tests for the transfers in parallel and replaced the coarse grid solver in the parallel case by a PCG solver preconditioned with AMG. This results in robust iteration numbers in parallel. I checked it on the CPU for up to 8 MPI ranks.

Unfortunately, the transfers are not yet 100% matrix-free because of the call to `MultTranspose` of the parallel restriction matrix in [transfer.cpp#L559](https://github.com/mfem/mfem/blob/drzisga/multigrid/fem/transfer.cpp#L559). The matrix-free implementation of this will require some communication which is not straightforward for me to implement. Is someone maybe already working on this? The call to `GetProlongationMatrix` already returns matrix-free operators and I would need the equivalent matrix-free operators for `GetRestrictionMatrix`.

v-dobrev commented:
@drzisga, the restriction operator is local and represented by a `SparseMatrix` which has at most 1 entry per row. This is not optimal but it is not that bad.

drzisga commented:
Thank you for the input @v-dobrev. I changed it back and use the RestrictionMatrix again. I also tested it on a single GPU and get the same results back. @pazner can you please test it again?

pazner commented:
Thanks @drzisga! I'll take another look at it

barker29 commented on tests/unit/fem/test_transfer.cpp:
```diff
@@ -0,0 +1,311 @@
+                  Mesh* mesh;
+                  if (dimension == 2)
+                  {
+                     mesh = new Mesh(ne, ne, Element::QUADRILATERAL, 1, 1.0, 1.0);
```
I'm probably nitpicking, but would it be easy to add a test for tris/tets? The p-refinement in particular has a different code path in this case that I don't think is unit tested at the moment.

drzisga replied:
It was quite easy to add. Triangles and tetrahedra are being tested now too.

tzanio commented:
@delyank, can you also take a quick look and re-confirm if you approve?

pazner commented:
Do you know why the Mac runs on Travis hung?

drzisga commented on examples/ex25.cpp:
```diff
@@ -0,0 +1,278 @@
+   //     using GLVis: "glvis -m refined.mesh -g sol.gf".
+   ofstream mesh_ofs("refined.mesh");
+   mesh_ofs.precision(8);
+   mesh->Print(mesh_ofs);
```
Thank you for noticing this. I fixed it here and at the other locations.

drzisga commented on fem/transfer.cpp:
```diff
@@ -514,30 +514,38 @@ void TensorProductPRefinementTransferOperator::MultTranspose(const Vector& x,
+#ifdef MFEM_USE_MPI
+TrueTransferOperator::TrueTransferOperator(const ParFiniteElementSpace&
+                                           lFESpace_,
+                                           const ParFiniteElementSpace& hFESpace_)
```
Make style messed this up. I fixed it manually.

drzisga commented:
> Do you know why the Mac runs on Travis hung?

It seems to be a Travis issue or am I wrong?

tzanio commented:
> > Do you know why the Mac runs on Travis hung?
> 
> It seems to be a Travis issue or am I wrong?

The Mac runs on Travis have been taking forever the last week. We try to restart them, but the issue seems to definitely be on the Travis end...

tzanio commented:
LGTM

tzanio commented:
Merged in `next` for testing...

tzanio commented:
Reviewers: do the results from tonight's (2020-03-24) regression testing look ok to everyone?

1. [tux429](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/de7c73ba1a84cbe693f93a503d8aa76dcecc4c4d#tux429/2020-03-24-github-next/sample-runs-devices-tux429.diff)
2. [tux375](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/8d65dbd761991b80662a1f2b8e7126690e05f7c7#tux375/2020-03-24-llnl-gh-next/sample-runs-baseline-tux375.diff) (you need to clone it locally to view)

barker29 commented:
The device runs on tux429 look great. For the baselines on tux375, I am seeing a few places where residuals change more than I expect (around `10^{-6}`) but I am quite sure those differences don't come from this PR.

tzanio commented:
Thanks Andrew

drzisga commented:
I am looking into this.

drzisga commented:
I could trace it back to the changes I have made in the `RefinementOperator::Mult` method [here](https://github.com/mfem/mfem/blob/drzisga/multigrid/fem/fespace.cpp#L1104). I got rid of the `processed` array and use the `SetSubVector` method instead which just overwrites previously processed values. Usually, this should make no difference, but rounding errors can change things. I don't understand, however, why this makes such a large difference of `10^{-6}`. Could this be due to the sensitivity of Krylov solvers to rounding errors?

Should I revert this method back to the version from master?

tzanio commented:
@drzisga -- what was the reason for the above change, efficiency, clarity, something else?

tzanio commented:
I also have a more "philosophical" design question...

If I understand correctly, the new classes introduced can be seen as multilevel versions of the main building blocks of Example 1:

```c++
FiniteElementSpace -> SpaceHierarchy
BilinearForm/OperatorPtr -> MultigridOperator
GSSmoother/Solver -> MultigridSolver
```

Is that correct?

From that perspective, I think it will be better to

1) Rename `MultigridOperator` to `OperatorHierarchy` and for it to be a stand-alone class, not inherit from Operator (just like `SpaceHierarchy` does not inherit from `FiniteElementSpace`).

2) To keep things simple, you may also consider to use plural instead of "Hierarchy", e.g. `SpaceHierarchy -> Spaces`, `OperatorHierarchy -> Operators`.

3) Rename `MultigridSolver` to just `Multigrid`

The code of ex26 then can then mimic ex1 closer:

```c++
FiniteElementSpace *fespace = new FiniteElementSpace(mesh, fec);
LinearForm(fespace);
BilinearForm *a = new BilinearForm(fespace);
a->AddDomainIntegrator(new DiffusionIntegrator(one));
OperatorPtr A;
Vector B, X;
a->FormLinearSystem(ess_tdof_list, x, *b, A, X, B);
GSSmoother M((SparseMatrix&)(*A));
PCG(*A, M, B, X, 1, 200, 1e-12, 0.0);
```

```c++
Spaces fespaces(mesh, fespace, true, true);
LinearForm *b = new LinearForm(&fespaces.GetFinest());
DiffusionOperators A(fespaces, ess_bdr);
Vector B, X;
A.GetFinest().EliminateBCs(x, *b, X, B);
Multigrid M(&A, Multigrid::CycleType::VCYCLE, 1, 1);
PCG(A.GetFinest(), M, B, X, 1, 2000, 1e-12, 0.0);
```

What do you think?

v-dobrev commented:
I also looked at this and tried reverting the changes in `RefinementOperator::Mult`, here: https://github.com/mfem/mfem/blob/b5655c5f56cfc556b1aa009b9703781a85e1b03f/fem/fespace.cpp#L1103-L1138 to the version in `master` here: https://github.com/mfem/mfem/blob/9431806aa73917545404d70cb24017da7dc6a238/fem/fespace.cpp#L1093-L1141 That made at least some of the differences disappear (I did not run all the tests).

I think it maybe better to keep the version from `master` which, if I remember correctly, was optimized to some extend by @jakubcerveny. @jakubcerveny, do I remember correctly?

@drzisga, was this modification done just to simplify implementation, or you had some other reason? If it was just for simplification, it maybe better to keep the original version if that implementation is a bit faster.

jakubcerveny commented:
I like the new version of `RefinementOperator::Mult`, and I think it used to look like that in the past but then we had to expand the multiplication and do it by explicitly element by element. But I can't remember the reason for that and why it is actually better. It would be nice to figure this out and keep the concise version.

drzisga commented:
@v-dobrev My intention was to get rid of the `processed` array which has the same size as the vector. My goal was to reduce memory consumption and traffic, since no additional data needs to be read from memory. However, I have not benchmarked the code and cannot say if it got actually better or worse.

@jakubcerveny maybe it had something to do with the decoding and encoding of the dofs? The `GetSubVector` and `SetSubVector` methods do this internally.

drzisga commented:
@tzanio I like your idea very much, since that was my original intention. The only thing that does not really fit in this case is, that the smoothers are included in `OperatorHierarchy` or `Operators` and not in the `Multigrid` class. This is somehow related to @barker29's comment a few days ago, where he asked why I do not have separate classes for them. I will rethink this.

tzanio commented:
> The only thing that does not really fit in this case is, that the smoothers are included in OperatorHierarchy or Operators and not in the Multigrid class. This is somehow related to @barker29's comment a few days ago, where he asked why I do not have separate classes for them. I will rethink this.

Can't the smoothers be part of the Multigrid class? That is most natural, because the type of smoother is one of the multigrid properties and there could be different smoothing options that the user can set...

drzisga commented:
This would easily work for matrix-based operators where the smoother can access the matrix diagonal. In the partial assembly case, the smoother needs to obtain the assembled diagonal from the `BilinearForm`. My goal was to keep the `MultigridSolver` a pure `linalg` class without knowing anything of finite elements or bilinear forms. This is why I moved the smoothers into the `MultigridOperator`, but I am not perfectly happy about that.

barker29 commented:
@tzanio That was exactly the discussion we had several days ago. Probably either option is possible, but there's always tradeoffs. @drzisga chose one option, the reviewers agreed to it, it got merged in next, and now you're rewriting the PR because you ended up looking at it for some unrelated reason. I don't have strong feelings about the design decision itself but from a workflow perspective this is suboptimal.

pazner commented:
> I don't have strong feelings about the design decision itself but from a workflow perspective this is suboptimal.

I agree that it is a bit late in the game to have these discussions, but I will take some of the blame because I brought it up in our PR discussion yesterday.

I think this multigrid functionality could become widely used by MFEM users, so it's worth it to come up with an interface that everyone is happy with before it's too late to make interface changes (i.e. merged into master).

pazner commented:
Another option would be to unify the `MultigridOperator` and `MultigridSolver` classes. Is the reason to have them separate so that multiple cycles can be defined on the same operator hierarchy? I don't think that will be a very common use case.

Currently, `MultigridOperator` doesn't make sense as a class on its own without an accompanying `MultigridSolver`. If both classes always will have to be defined, I think it makes sense to combine them.

tzanio commented:
> @drzisga chose one option, the reviewers agreed to it, it got merged in next, and now you're rewriting the PR because you ended up looking at it for some unrelated reason.  I don't have strong feelings about the design decision itself but from a workflow perspective this is suboptimal.

@barker29, I apologize for not getting involved earlier.

I didn't realize that the interface has issues before they were raised at the PR discussion meeting yesterday. I very much appreciate the work you @delyank and @pazner did reviewing this PR, and ordinarily I would have not jumped in at this point. 

But this is a special case -- I agree with @pazner that Multigrid will be widely used and so the user-facing design is particularly important (maybe even more important than the implementation, which could later change). My intention is not to rewrite the PR, I am just trying to suggest minimal tweaks that could make the interface more clear and logical to users.

v-dobrev commented:
I think it is best to address any remaining concerns before merging.

Let's discuss this in a telecon and decide what to do. Some of us are focused on the CEED milestone at the moment, so next week is probably a better time for the telecon. I'll send out a telecon proposal by tomorrow.

barker29 commented:
I agree that we should address this, and I apologize for my earlier annoyed comments.

delyank commented:
> It's looking like the differences on tux375 are caused by this PR. We can see the differences by comparing the residual at iteration 1 of the run (comparing master to this PR):
> 
> ```
> ./ex8 -m ../data/square-disc-p3.mesh -no-vis
> ```
> 
> We should look into what's causing this.

I don't know if I'm looking at the correct thing, but what I see doesn't seem very bad. Yes, there is difference in the first residual but it's relatively OK. Actually, the number coinciding significant digits is better the other residuals. I don't think that difference in the residuals is so indicative, as long as the residuals are of a similar magnitude.

v-dobrev commented:
I merged `master` here (to resolve conflicts).

Also, for testing purposes, I copied the implementation of `RefinementOperator::Mult` from `mastrer` while keeping the new implementation commented out. I'll re-merge this in `next` in a moment -- that will allow us to see if there are any differences in the autotests that are due to other changes from this PR. We can decide which version to keep at the telecon.

v-dobrev commented:
Re-merged in `next` for testing.

tzanio commented:
Mockup from 4/1/20 telecon:
```c++
FiniteElementCollection *fec = new H1_FECollection(1, dim);
FiniteElementSpace *coarse_fespace = new FiniteElementSpace(coarse_mesh, fec);
FiniteElementSpaceHierarchy fespaces(mesh, coarse_fespace, true, true);
LinearForm *b = new LinearForm(&fespaces.GetFineSpace());
DiffusionMultigrid M(fespaces, ess_bdr);
M.SetCycleType(Multigrid::VCYCLE, 1, 1);
OperatorPtr A;
Vector B, X;
M.FormFineLinearSystem(x, *b, A, X, B);
PCG(*A, M, B, X, 1, 2000, 1e-12, 0.0);
```

delyank commented:
> Mockup from 4/1/20 telecon:
> 
> ```c++
> FiniteElementCollection *fec = new H1_FECollection(1, dim);
> FiniteElementSpace *coarse_fespace = new FiniteElementSpace(coarse_mesh, fec);
> FiniteElementSpaceHierarchy fespaces(mesh, coarse_fespace, true, true);
> // ...
> LinearForm *b = new LinearForm(&fespaces.GetFineSpace());
> DiffusionMultigrid M(fespaces, ess_bdr);
> M.SetCycleType(Multigrid::VCYCLE, 1, 1);
> OperatorPtr A;
> Vector B, X;
> M.FormFineLinearSystem(x, *b, A, X, B);
> PCG(*A, M, B, X, 1, 2000, 1e-12, 0.0);
> ```
> 
> Other decisions:
> 
>     * Consider keeping `TransferOperator`s inside `FiniteElementSpaceHierarchy`
> 
>     * Switch to the new implementation of `RefinementOperator::Mult()`
> 
>     * Try to move as much as possible of the default functionality to `Multigrid` so the derived classes, such as `DiffusionMultigrid` are as small as possible and code duplication is avoided

These seem like not too large changes and I appreciate the ideas. So it will hopefully allow merging the PR soon, as it is a quite good PR.

drzisga commented:
Thank you all again for the feedback. I have now added the following changes we discussed:
- Switched to my implementation of `RefinementOperator::Mult()`
- Renamed `SpaceHierarchy` to `FiniteElementSpaceHierarchy`
- Moved `TransferOperator`s inside `FiniteElementSpaceHierarchy`. The `FiniteElementSpaceHierarchy` now manages ownership of the prolongations.
- Moved `MultigridSolver` into `Multigrid`
- I tried to move as much as possible of the default functionality to `Multigrid`
- Reduced code duplication in `DiffusionMultigrid`
- Changed the _interface_ according to the mockup from the telecon

tzanio commented:
> Thank you all again for the feedback. I have now added the following changes we discussed:
> 
> * Switched to my implementation of `RefinementOperator::Mult()`
> * Renamed `SpaceHierarchy` to `FiniteElementSpaceHierarchy`
> * Moved `TransferOperator`s inside `FiniteElementSpaceHierarchy`. The `FiniteElementSpaceHierarchy` now manages ownership of the prolongations.
> * Moved `MultigridSolver` into `Multigrid`
> * I tried to move as much as possible of the default functionality to `Multigrid`
> * Reduced code duplication in `DiffusionMultigrid`
> * Changed the _interface_ according to the mockup from the telecon

Thanks for addressing these @drzisga, the interface now looks good to me!

tzanio commented on examples/ex26.cpp:
```diff
@@ -0,0 +1,243 @@
+using namespace std;
+using namespace mfem;
+
+class DiffusionMultigrid : public Multigrid
```
Add a short description

drzisga replied:
I added a description here and in ex26p

tzanio commented on examples/ex26.cpp:
```diff
@@ -0,0 +1,243 @@
+   }
+};
+
+int main(int argc, char *argv[])
```
Two lines between the class and `main()`?

drzisga replied:
I added a newlines here and in ex26p

tzanio commented on examples/ex26.cpp:
```diff
@@ -0,0 +1,243 @@
+   //    FiniteElementSpaceHierarchy and additional boundary information. This operator
+   //    is then used to create the MultigridSolver as a preconditioner in the
+   //    iterative solver.
+   Array<int> ess_bdr(mesh->bdr_attributes.Max());
+   ess_bdr = 1;
+   DiffusionMultigrid M(fespaces, ess_bdr);
+   M.SetCycleType(Multigrid::CycleType::VCYCLE, 1, 1);
+   OperatorPtr A;
+   Vector B, X;
+   M.FormFineLinearSystem(x, *b, A, X, B);
+   cout << "Size of linear system: " << A->Height() << endl;
```
```suggestion
   Array<int> ess_bdr(mesh->bdr_attributes.Max());
   ess_bdr = 1;

   DiffusionMultigrid M(fespaces, ess_bdr);
   M.SetCycleType(Multigrid::CycleType::VCYCLE, 1, 1);

   OperatorPtr A;
   Vector B, X;
   M.FormFineLinearSystem(x, *b, A, X, B);
   cout << "Size of linear system: " << A->Height() << endl;
```

tzanio commented on linalg/eigenvalue.hpp:
```diff
@@ -0,0 +1,99 @@
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
```
I wonder if this will be better in `solvers.?pp`... what do the reviewers think?

delyank replied:
In this form, it "lives" in its own world. It doesn't inherit from the `Solver` class and does not follow any pre-existent interface. On the other hand, it is in a way a "solver".

Maybe it can be put in the terms of the `Solver::Mult(...)` interface and return also the eigenvector, but I'm not a big fan of that idea.

If we look at already existing eigenslover, they seem to be in the matrix files and classes. Notably, the class `DenseMatrixEigensystem` is in `densemat.*`. So if we follow that paradigm, this needs to move to `operator.*`

tzanio replied:
I don't think strongly about this either way...

tzanio replied:
There are eigensolvers also in hypre and the densemat files

delyank replied:
Actually, yes, I'm for moving this to `operator.*`, but I also don't have strong opinion.

drzisga replied:
I moved the `PowerMethod` class to `operator.*`

tzanio commented:
LGTM

tzanio commented:
Re-merged in `next` for testing ...

v-dobrev commented:
Re-merged in `next`.

v-dobrev commented on fem/bilinearform.hpp:
```diff
@@ -151,8 +151,8 @@ class BilinearForm : public Matrix
 
-   /// Get the assembly level
-   AssemblyLevel GetAssemblyLevel() {return assembly;}
+   /** Returns the assembly level */
+   AssemblyLevel GetAssemblyLevel() const { return assembly; }
```
There is a difference between `/// ...` and `/** ... */` doxygen comments -- we probably should add a section about this in `CONTRIBUTING.md`.

In this case, we want the comment to become a "brief" documentation and since it fits on one line we can use `/// ...` comment. Alternatively, (or if the `/// ...` comment does not fit on one line) one can use `/** @brief ... */` comment to create a "brief" documentation. The comment `/** ... */` without `@brief` becomes a "long" documentation (and so does a multiline `/// ...` comment without `@brief`).

The "brief" documentation is shown next to the method when listed as part of all class methods. The "long" documentation (together with the "brief" documentation) is shown in the detailed method description which can be accessed by clicking the link "More ..." that is added by doxygen at the end of the "brief" documentation.

Doxygen has a number of such conventions that are not obvious, so it is always best to run `make doc` and look at the generated output, e.g. by opening `doc/CodeDocumentation.html` and then searching for the relevant class through the search box in the top right corner.

drzisga replied:
Thank you very much for the information @v-dobrev. I fixed the documentation according to your comments.
